[STEP 3-1-1] 만료시간 추가

jiheon2234 opened the pull request:
- api 사용자가 단축 url을 만들 때, ttlMinutes를 추가할 수 있도록 변경
- 원래는 단축된 URL만 저장했으나, 요구사항의 변경에 따라 ID, 원본URL, 단축URL, expiredAt 을 한번에 저장하도록 변경
- 만료시간이 지난 URL은 조회할 수 없도록 변경 (추후에 스케줄링을 통한 논리적 삭제로 변경 예정)

jiheon2234 committed:
feature : 만료시간 추가
- api 사용자가 단축 url을 만들 때, ttlMinutes를 추가할 수 있도록 변경
- 원래는 단축된 URL만 저장했으나, 요구사항의 변경에 따라 ID, 원본URL, 단축URL, expiredAt 을 한번에 저장하도록 변경
- 만료시간이 지난 URL은 조회할 수 없도록 변경 (추후에 스케줄링을 통한 논리적 삭제로 변경 예정)

jiheon2234 commented on src/main/java/community/whatever/onembackendjava/repository/URLShortenRepository.java:
```diff
@@ -2,22 +2,23 @@
 import java.util.Optional;
 
+import community.whatever.onembackendjava.entity.ShortenedURLEntity;
+
```
단순히 단축 URL(String)만 저장하는 것이 아니라, 관련된 정보를 함께 저장하도록 변경하였습니다.

jiheon2234 commented on src/main/java/community/whatever/onembackendjava/entity/ShortenedURLEntity.java:
```diff
@@ -0,0 +1,32 @@
+import lombok.Getter;
+
+@Getter
+public class ShortenedURLEntity {
```
원본 url, 단축 url, 만료시간, 삭제여부를 가지고 있는 엔티티입니다.

jiheon2234 commented on src/main/java/community/whatever/onembackendjava/service/ShortenURLService.java:
```diff
@@ -29,28 +33,35 @@ public class ShortenURLService {
-	public String createShortenedURL(String originURL) {
-		if (checkDomainInBlackList(originURL)) {
+	public ShortenedURLCreateResponse createShortenedURL(ShortenedURLCreateRequest req) {
+		if (checkDomainInBlackList(req.originURL())) {
```
만료시간이 지났으면 응답이 반환되지 안도록 설정하였습니다.
추후에 스케줄러로 변경 예정입니다.

jiheon2234 commented on src/main/java/community/whatever/onembackendjava/service/ShortenURLService.java:
```diff
@@ -29,28 +33,35 @@ public class ShortenURLService {
+			.originURL(req.originURL())
+			.shortenedURL(generatedShortenedURL)
+			.expiredAt(getExpirationTime(req.ttlMinutes()))
+			.build();
```
테스트

jiheon2234 commented on src/main/java/community/whatever/onembackendjava/entity/ShortenedURLEntity.java:
기존에 단축된 url string값만 저장하는대신, 연관된 정보들을 함께 저장하도록 변경하였습니다.

jiheon2234 commented on src/main/java/community/whatever/onembackendjava/service/ShortenURLService.java:
```diff
@@ -29,28 +33,35 @@ public class ShortenURLService {
-		String originURL = repository.findByShortenedURL(shortenedURL)
+		ShortenedURLEntity entity = repository.findByShortenedURL(shortenedURL)
+			.filter((se) -> se.getExpiredAt().isAfter(LocalDateTime.now()))
 			.orElseThrow(() -> BusinessLogicException.from(BusinessExceptionCode.ORIGIN_URL_NOT_FOUND));
```
만료시간이 지난 URL들은 조회시 응답하지 않도록 설정했습니다.
추후 스케줄러를 통한 논리적 삭제 방식으로 변경할 생각입니다.

Hyune-c commented on src/main/java/community/whatever/onembackendjava/dto/req/ShortenedURLCreateRequest.java:
```diff
@@ -1,10 +1,16 @@
+
+	@Min(value = 0, message = "만료 기간은 음수일 수 없습니다.")
+	@Max(value = 1440, message = "만료 기간은 최대 하루(1440분) 입니다.")
+	int ttlMinutes
```
`nit` wrapper 자료형 사용을 권장 합니다.

Hyune-c commented on src/main/java/community/whatever/onembackendjava/entity/ShortenedURLEntity.java:
```diff
@@ -0,0 +1,32 @@
+		this.originURL = originURL;
+		this.shortenedURL = shortenedURL;
+		this.expiredAt = expiredAt;
+	}
```
빌더는 위험 합니다. 불가피하다면 생성자나 팩토리 메서드 안에서만 사용을 권장 합니다.
- 위험한 이유는 직접 검색해보세요.

jiheon2234 replied:
추후 스태틱 생성자 변경 예정입니다!

Hyune-c commented on src/main/java/community/whatever/onembackendjava/entity/ShortenedURLEntity.java:
```diff
@@ -0,0 +1,32 @@
+
+	public void markDelete() {
+		this.isDeleted = true;
+	}
```
fyi. disable, expiry

jiheon2234 replied:
지금같은 경우에는expiry가 훨씬 직관적인 네이밍인 것 같습니다.

다만  만약 테이블이 많아지고, 모두 soft delete 한다면
테이블의 경우마다 알맞게 "삭제" 메서드명이 통일되지 않아 혼란스러울 것 같습니다.

약간 주제를 벗어나는 내용인데, 실무에서 jpa를 사용하실때
이러한 논리적 삭제를 @sqldelete를 사용하는지, 아니면 명시적으로 markDelete()를 호출하고 변경감지를 이용하는지 궁금합니다. (제 생각에는 markDelete를 호출하는 쪽이 좀 더 객체지향적인것 같습니다. 삭제와 함께 연관되는 작업들이 있을 수도 있고....)
추가로 @sqlRestriction(@where)등의 어노테이션을 사용하는지, 아니면 개발자가 조건을 명시하도록 하는지도 궁금합니다. (이건 당연히 요구사항에 따라 다르고 장단점이 분명하지만 실무에서 많이 쓰이는 조합이 궁금합니다.!)

Hyune-c replied:
> 실무에서 jpa를 사용하실때
이러한 논리적 삭제를 @sqldelete를 사용하는지, 아니면 명시적으로 markDelete()를 호출하고 변경감지를 이용하는지 궁금합니다. 

저 `@sqldelete` 가 있는 것도 처음 알았어요 =_=;;
찾아보니까 jpql 정의하고 실행하는 거네요. 앞으로도 안쓸거 같습니다.
- soft delete 가 같은 테이블에서 플래그만 바꾸는 것인가?
- 테이블 row 수 간결화를 위해 delete 대상은 archives 테이블로 이관되어야 한다면?
- delete 대상이 외부 서비스로 전파되어야 한다면?
- 영속화 대상이 rdb 가 아니게 된다면?
- 이 모든게 다 일어나야 된다면?

jiheon2234 committed:
지적받은사항 수정

- ShortenedURLCreateRequest.ttlMinutes Int->Integer
- ShortenedURLEntity.isDeleted -> disabled로 변경
- 빌더 대신 스태틱 생성자를 사용하게 변경
